torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -855,20 +917,44 @@ class ShapePropagator {
         [this](Node* node) -> type_vec_t {
           if (auto maybe_tensor_types =
                   gatherTensorTypes<DimensionedTensorType>(node)) {
-            return {broadcast(*maybe_tensor_types, 0)};
+            return {broadcast(*maybe_tensor_types, (*maybe_tensor_types)[0]->scalarType())};
           }
           return {};
         }};
 
-    // NB: we always take the scalar type of the Tensor
-    static const register_formula_for broadcasting_tensor_scalar_ops{
+    static const register_formula_for broadcasting_tensor_scalar_ops_arithmetic{
         {
-
             // Tensor-Scalar operators
             "aten::add(Tensor self, Scalar other, Scalar alpha) -> Tensor",
             "aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor",
             "aten::mul(Tensor self, Scalar other) -> Tensor",
             "aten::div(Tensor self, Scalar other) -> Tensor",
+          },
+          [this](Node* node) -> type_vec_t {
+            if (auto maybe_tensor_types =
+                    gatherTensorTypes<DimensionedTensorType>(node)) {
+              auto first_scalar_type = (*maybe_tensor_types)[0]->scalarType();
+              auto second_scalar_type = typeToScalarType(node->inputs()[1]->type());
+              if (isIntegralType(first_scalar_type) && isFloatingType(second_scalar_type) )
```

Comment:
The main thing I was trying to address on this commit was we were going to differentiate signed/unsigned as categories, but now differentiate bool/non-bool instead.

"if you don't know its concrete type" ... I'm not clear what you mean by that? if `Scalar other` is a float or int won't the typeToScalarType / scalarTypeFromJitType return an int or float? What's the case where we don't know the type? When calling torch.add(), both arguments have to have types of some sort don't they?

`typeToScalarType` no longer returns an optional, so `other` can't be an optional/nullopt.

Reply:
There are a lot of terms here with overloaded meanings so i understand the confusion.
`Scalar` in our schema refers to a union of IntType || FloatType. The input will either be an IntType, a FloatType, or the NumberType itself, in which case the runtime type will either be an int or a float. 
In the following case:
```
de foo(x):
    scalar_number = x.item()
    return torch.tensor(1).add(other=scalar_number)
```
you have to give up on setting the dtype of the returned value because other will be NumberType and depends on the runtime dtype of `x`. (you could separately use the dtype of `x` to set `scalar_number` to its concrete type but that's out of scope for here).

Reply:
`getScalarType` should return an optional in the case that the input isn't a Dimensioned Tensor, int, float, or bool, to account for cases like that

Reply:
changed to return optional, added tests to capture the described cases